### PR TITLE
Added exponential backoff and jitter to CircuitBreaker

### PR DESCRIPTION
**Pull request by jhartlaub**

### CODE REVIEW

**Comment by renstrom** [comment]
Thanks for the PR. You're gonna have to fix those PEP8/Frosted violations first though. You can run `make check` locally, you need to have [pep8](https://pypi.python.org/pypi/pep8) and [frosted](https://pypi.python.org/pypi/frosted) installed though.

File: circuit/breaker.py [hunk 1]
```diff
@@ -44,8 +53,7 @@ class CircuitOpenError(Exception):
 class CircuitBreaker(object):
     """A single circuit with breaker logic."""
 
-    def __init__(self, clock, log, error_types, maxfail, reset_timeout,
-                 time_unit):
+    def __init__(self, clock, log, error_types, maxfail, reset_timeout, time_unit, backoff_cap=None, with_jitter=False):
```
Comment (renstrom) [inline circuit/breaker.py]:
```diff
-    def __init__(self, clock, log, error_types, maxfail, reset_timeout, time_unit, backoff_cap=None, with_jitter=False):
+    def __init__(self, clock, log, error_types, maxfail, reset_timeout,
+                 time_unit, backoff_cap=None, with_jitter=False):
```

File: circuit/breaker.py [hunk 2]
```diff
@@ -127,12 +144,14 @@ class CircuitBreakerSet(object):
     """
 
     def __init__(self, clock, log, maxfail=3, reset_timeout=10,
-                 time_unit=60, factory=CircuitBreaker):
+                 time_unit=60, backoff_cap=None, with_jitter=False, factory=CircuitBreaker):
```
Comment (renstrom) [inline circuit/breaker.py]: Put `factory=CircuitBreaker` on a third row to comply with PEP8.

```python
    def __init__(self, clock, log, maxfail=3, reset_timeout=10,
                 time_unit=60, backoff_cap=None, with_jitter=False,
                 factory=CircuitBreaker):
```

File: circuit/breaker.py [hunk 3]
```diff
@@ -85,7 +99,10 @@ def test(self):
         """
         if self.state == 'open':
             delta = self.clock() - self.last_change
-            if delta < self.reset_timeout:
+            delay_time = self.backoff_cap and min(self.reset_timeout * (2 ** self.test_fail_count), self.backoff_cap) or self.reset_timeout
```
Comment (renstrom) [inline circuit/breaker.py]: This line is a bit hard to crack, might have to get a bit creative with intermediary variable. Something like:

```python
delay_time = self.reset_timeout
if self.backoff_cap:
    delay_time = self.reset_timeout * (2 ** self.test_fail_count)
    delay_time = min(delay_time, self.backoff_cap)
```

File: circuit/breaker.py [hunk 4]
```diff
@@ -85,7 +99,10 @@ def test(self):
...
+            if self.with_jitter:
+                delay_time = random.random() * delay_time  # add jitter, see http://www.awsarchitectureblog.com/2015/03/backoff.html
```
Comment (renstrom) [inline circuit/breaker.py]:
```diff
-                delay_time = random.random() * delay_time  # add jitter, see http://www.awsarchitectureblog.com/2015/03/backoff.html
+                # Add jitter, see:
+                # http://www.awsarchitectureblog.com/2015/03/backoff.html
+                delay_time = random.random() * delay_time
```

**Comment by renstrom** [comment]
@bergman should we just merge this and fix the formatting issues ourselves? What do you think?

**Comment by bergman** [comment]
@renstrom sure, or just fix it in the merge commit (don't use the merge button on github).

**Pull request merged**